src/Function.cpp

```diff
@@ -248,6 +338,103 @@ Function::Function(const std::string &n) : contents(new FunctionContents) {
     contents.ptr->name = n;
 }
 
+void deep_copy_update_definition_helper(IntrusivePtr<FunctionContents> &contents,
+                                        UpdateDefinition &dst,
+                                        const UpdateDefinition &src,
+                                        DeepCopyMap &copied_map) {
+    dst.values = src.values;
+    dst.args = src.args;
+    dst.schedule = Schedule();
+    src.schedule.deep_copy(dst.schedule, copied_map);
+    // UpdateDefinition's domain is the same as the one pointed by its schedule
+    internal_assert(src.schedule.reduction_domain().same_as(src.domain))
+        << "UpdateDefinition should point to the same reduction domain as its schedule\n";
+    dst.domain = dst.schedule.reduction_domain();
+}
+
+void deep_copy_extern_func_argument_helper(ExternFuncArgument &dst,
+                                           const ExternFuncArgument &src,
+                                           DeepCopyMap &copied_map) {
+    dst.arg_type = src.arg_type;
+    dst.buffer = src.buffer;
+    dst.expr = src.expr;
+    dst.image_param = src.image_param;
+    if (copied_map.count(src.func)) {
```

Comment:
It's a pretty minor thing, but I always hate it when I'm forced to make two map lookups like this. Maybe something like this would work?

    IntrusivePtr<FunctionContents> &copied_func = copied_map[src.func];
    if (copied_func != nullptr) {
        dst.func = copied_func;
    } else {
        dst.func = IntrusivePtr<FunctionContents>(new FunctionContents);
        ...
        copied = dst.func;
    }